### PR TITLE
Add remove(where:) to RangeReplaceableCollection

### PR DESCRIPTION
**Pull request by airspeedswift**
Initial implementation for https://github.com/apple/swift-evolution/pull/675

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test linux platform

File: stdlib/public/core/RangeReplaceableCollection.swift.gyb [hunk 1]
```diff
@@ -509,6 +509,19 @@ public protocol RangeReplaceableCollection
   ///
   /// - Complexity: O(*n*), where *n* is the length of the collection.
   mutating func removeAll(keepingCapacity keepCapacity: Bool /*= false*/)
+
+  /// Removes all elements from the collection.
```
Comment (dabrahams) [inline stdlib/public/core/RangeReplaceableCollection.swift.gyb]: This summary seems to be wrong..

Given that the names in the proposal have been changed, and the intent behind having an implementation is that people can get actual experience with the proposal, I think this PR should be updated before we run the review.

File: stdlib/public/core/RangeReplaceableCollection.swift.gyb [hunk 2]
```diff
@@ -509,6 +509,19 @@ public protocol RangeReplaceableCollection
...
+  ///
+  /// Calling this method may invalidate any existing indices for use with this
+  /// collection.
+  ///
+  /// - Parameter keepCapacity: Pass `true` to request that the collection
```
Comment (dabrahams) [inline stdlib/public/core/RangeReplaceableCollection.swift.gyb]: This parameter doesn't seem to exist

File: stdlib/public/core/RangeReplaceableCollection.swift.gyb [hunk 3]
```diff
@@ -1226,7 +1239,7 @@ extension RangeReplaceableCollection {
   ///
   /// - Parameter isIncluded: A closure that takes an element of the
   ///   sequence as its argument and returns a Boolean value indicating
-  ///   whether the element should be included in the returned array.
+  ///   whether the element should be included in the returned collection.
```
Comment (dabrahams) [inline stdlib/public/core/RangeReplaceableCollection.swift.gyb]: You might want to make that fix separately from this PR

File: stdlib/public/core/RangeReplaceableCollection.swift.gyb [hunk 4]
```diff
@@ -1237,6 +1250,67 @@ extension RangeReplaceableCollection {
   }
 }
 
+extension RangeReplaceableCollection where Self: MutableCollection {
+  /// Removes from the collection all elements that satisfy the given predicate.
+  ///
+  /// - Parameter where: A closure that takes an element of the
```
Comment (dabrahams) [inline stdlib/public/core/RangeReplaceableCollection.swift.gyb]: The parameter name is `predicate`, not `where`, thank goodness!

File: test/stdlib/RemoveElements.swift [hunk 5]
```diff
@@ -0,0 +1,101 @@
+//===--- Filter.swift - tests for lazy filtering --------------------------===//
```
Comment (dabrahams) [inline test/stdlib/RemoveElements.swift]: Copy/paste error?

Comment (airspeedswift) [reply test/stdlib/RemoveElements.swift]: derp... thx

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by martinr448** [comment]
I am not sure if it matters, but in `remove(where:)` the predicate is always evaluated *twice* for the first matching element, because the starting value for `j` is `i`, not its successor.

**Comment by airspeedswift** [comment]
@martinr448 yeah, it also shouldn't use `indices`. I have been tinkering with this for a few different reasons including what the state should be if the predicate throws halfway, will push a new version soon.

**Pull request merged**